### PR TITLE
persist: inspect tool for printing contents of a batch part

### PR DESCRIPTION
**Pull request by pH14**
Wrote this up to debug an issue. Lets us inspect the contents of a particular batch part to see what's going on... check out the inline desc, what keys we're working over, etc.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist-client/src/inspect.rs [hunk 1]
```diff
@@ -201,6 +205,67 @@ pub async fn fetch_state_diffs(
...
+    let part = BlobTraceBatchPart::<u64>::decode(&part).expect("decodable");
+    let desc = part.desc.clone();
+
+    let mut encoded_part = EncodedPart::new(&*key, part.desc.clone(), part);
+    let mut out = BatchPartOutput {
+        desc,
+        updates: Vec::new(),
+    };
+    while let Some((k, v, t, d)) = encoded_part.next() {
+        out.updates.push(BatchPartUpdate {
```
Comment (danhhz) [inline src/persist-client/src/inspect.rs:258]: let's put in a flag to limit how many updates are included, with some reasonable default (20?). ideally we'd want to pull them from both the beginning and the end somehow

**Comment by benesch** [comment]
Ping on this! It's come in handy on at least two investigations. Anything keeping us from merging this?

File: src/persist-client/examples/inspect.rs [hunk 2]
```diff
@@ -93,6 +96,50 @@ pub struct StateArgs {
...
+    /// Blob to use
+    ///
+    /// When connecting to a deployed environment's blob, the necessary connection glue must be in
+    /// place. e.g. for S3, sign into SSO, set AWS_PROFILE and AWS_REGION appropriately, with a blob
+    /// URI scoped to the environment's bucket prefix.
+    #[clap(long)]
+    blob_uri: String,
+
+    /// Number of updates to output. Default is unbounded.
+    #[clap(long, default_value = "18446744073709551615")]
```
Comment (danhhz) [inline src/persist-client/examples/inspect.rs:139]: let's make the default bounded :)

**Review by danhhz — COMMENTED** [review]
LGTM modulo the unbounded limit by default.

btw, not for this PR, mostly a note to myself: now that we have workspace-hack, there's no practical reason that we couldn't push most of `persistcli inspect/admin` into src/persist-client itself, which would make it lighter weight to add these sorts of things. I might toss that up real quick, but I'll wait until this merges so you don't have to deal with the rebase

**Review by danhhz — APPROVED** [review]

File: src/persist-client/examples/inspect.rs (hunk 2 above)
Comment (pH14) [reply src/persist-client/examples/inspect.rs:139]: I'm partial to unbounded, similar to how a SQL query returns everything unless a limit is passed in. if we want it to be bounded by default, IMO we have a duty to structure the output to show the total count / maybe offset so a user immediately knows they're not seeing the full view, e.g.

```json
{
  updates: [...],
  count: 100,
}
```

it's not hard to do, but, I'm not sure we want to spend cycles on it

Comment (danhhz) [reply src/persist-client/examples/inspect.rs:139]: ha, SQL as an example of good/desirable UX is not particularly convincing :P, but various unix cli tools also agree with the precedent. sure, let's start with unbounded and see where it goes

Comment (pH14) [reply src/persist-client/examples/inspect.rs:119]: > ha, SQL as an example of good/desirable UX is not particularly convincing :P

LOL fair point

**Pull request merged**